### PR TITLE
api-on-the-fly - part 4 - mutations

### PR DESCRIPTION
**Pull request by oliver-sanders**
Addresses the "executing mutations" point in #339 

**Note** - *Built atop #382 *
**Note** - *Works with the data types defined in cylc/cylc-flow#3469*

Add the ability to run mutations from the temporary Mutations view.

Wraps the mutations view up as a Lumino widget in the workflows view, still a a temporary measure but a good POC.

We *could* automatically fill in the workflow id field based on the current workflow but I think that'll have to wait for the next phase.

**TODO:**

* [x] The "Broadcast" mutation doesn't work due to an inconsistency between the server and schema arguments in Cylc Flow. - **Fixed in latest commit on the Cylc Flow PR**
* [x] Handling of null types needs to improve. - **Seems to work now with minor changes**
* [x] Tests currently broken awaiting resolution of the previous point.  - **Fixed**
* [x] Covered in debug information.  - **Sorted**

**What you should expect to see:**

1. Go to workflows
2. Open the mutations view.
3. Pick a mutation from the dropdown.
4. Fill in the form (note workflow id is `<user>|<reg>`).
5. Press submit.

**Example:**

From the UI:

aoft2

In the suite log:

```
2020-01-24T04:11:05Z INFO - Command succeeded:
	reset_task_states(['*.20191214T1200Z'], state=expired, outputs=[])
2020-01-24T04:11:05Z INFO - Processing 1 queued command(s)
	+       reset_task_states(['*.20191214T1200Z'], state=expired, outputs=[])
```

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] No change log entry required (why? e.g. temporary view not exposed to users).
- [x] No documentation update required.

### CODE REVIEW

**Comment by kinow** [comment]
I think this must be the last or penultimate PR for AOF in thr UI?? 🎉

**Comment by oliver-sanders** [comment]
Three bullet points left, but getting close. Auto-completing task names and cycle points will be fun.

**Comment by oliver-sanders** [comment]
Finally managed to compile peek after manually hacking some of the generated files to pick up the correct environment, something is jiggered in the recording/conversion, meh, screenshots will have to do for now.

**Comment by hjoliver** [comment]
I know the mutations view is temporary ... but using waiting and succeeded task status (name and icons) for the command status could be a bit confusing.  Maybe just changing the wording, something like "command pending" and "command succeeded"?

**Comment by kinow** [comment]
@oliver-sanders not working on infinite tree right now, so happy to review once the conflicts are solved 🎉

**Comment by hjoliver** [comment]
Conflicted.

**Comment by hjoliver** [comment]
Note the conflict resolution was not trivial, and I didn't entirely know what I was doing.  All in one commit in case @oliver-sanders needs to check it later 741f6f160030da249a944c4bee71963e67b26463

**Comment by kinow** [comment]
Sure thing @hjoliver, reviewing it right now.

File: QIntro [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                name
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
```
Comment (kinow) [inline QIntro]: Is this an example file @oliver-sanders ?

**Comment by kinow** [comment]
I think we are getting closer and closer to being able to call mutations. Tried to hold a running workflow, but I think the form is not ready to invoke that mutation yet (?). Got the following in the browser console:

image

But this PR adds a nice new widget, and keeps on improving the remaining code, so +1

**Comment by kinow** [comment]
> Note the conflict resolution was not trivial, and I didn't entirely know what I was doing. All in one commit in case @oliver-sanders needs to check it later [741f6f1](https://github.com/cylc/cylc-ui/commit/741f6f160030da249a944c4bee71963e67b26463)

Leaving final sign off to @oliver-sanders so that he can confirm the conflict resolution didn't miss anything, and that there's nothing that needs to be done after my comments :point_up:

File: src/utils/graphql.js [hunk 2]
```diff
@@ -120,3 +121,126 @@ export function createApolloClient (httpUrl, subscriptionClient) {
...
+ */
+export function getNullValue (type, types) {
+  if (!types) {
+    types = []
+  }
+  let ret = null
+  let ofType = null
+  for (const subType of iterateType(type)) {
+    if (subType.kind === 'LIST') {
+      ofType = getNullValue(subType.ofType)
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply src/utils/graphql.js]: That might be a conflict resolution fail on my part, I think.

File: QIntro (hunk 1 above)
Comment (hjoliver) [reply QIntro]: (Ha, I didn't even notice that :sweat_smile: )

**Comment by hjoliver** [comment]
> I think we are getting closer and closer to being able to call mutations. Tried to hold a running workflow, but I think the form is not ready to invoke that mutation yet (?).

That definitely worked for me - presumably you used the drop-down list at the top of the page (to choose "hold")?

**Comment by kinow** [comment]
> 
> 
> > I think we are getting closer and closer to being able to call mutations. Tried to hold a running workflow, but I think the form is not ready to invoke that mutation yet (?).
> 
> That definitely worked for me - presumably you used the drop-down list at the top of the page (to choose "hold")?

🤦‍♂ working as expected!

**Comment by hjoliver** [comment]
I guess my rushed conflict resolution attempt - in the hope of getting this in today - was a bit premature!

@oliver-sanders feel free to back out my commit if it's easier for you to do the resolution from scratch.  (It does test as working for me, but I think I kept almost every feature from both branches - not sure if that was intended - and I confess that some of the "example page" stuff apparent in @kinow 's screenshot above isn't clear to me).

**Comment by hjoliver** [comment]
Actually, after a bit of head-scratching from myself and @kinow above ... I know the "mutations view" is destined to  be a temporary thing, **but**:
- it's not clear what all the stuff on its front page is for, from "Sample Mutation" down; especially once you get down to the entries for `NonNull<List<NonNull<String>>> (required)` etc.
- all that stuff also obscures the drop-down list that gives access to all the real, individual mutations (It actually looks like merely a heading, with a small inverted triangle only visible off to the right)

File: QIntro (hunk 1 above)
Comment (oliver-sanders) [reply QIntro]: Yes, added by mistake, this BTW is a minimal hand-rolled introspection query.

Gotta admire the `ofType` recursive tree:

```graphql
  ofType {
    kind
    name
    ofType {
      kind
      name
      ofType {
```

**Comment by oliver-sanders** [comment]
> Tried to hold a running workflow, but I think the form is not ready to invoke that mutation yet (?)

Mutations should be working. Did you resolve this comment because you got it working? If not if you give me some steps to re-produce I'll take a crack at fixing it.

**Comment by hjoliver** [comment]
> The "Sample Mutation" is auto-generated from the schema ...

So, we should keep this in somehow (and documented) for use by developers in future.  (Development mode only?)

**Comment by oliver-sanders** [comment]
> So, we should keep this in somehow

Yep, somewhere, not quite sure where.

**Comment by oliver-sanders** [comment]
> Maybe my rushed conflict resolution attempt - in the hope of getting this in today - was a bit premature!

Wasn't the easiest merge ever, especially for a non-author! Found a couple of small errors so re-merged diffing against your attempt for validation.

File: src/components/cylc/Mutation.vue [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+// enumeration for the mutation status, maps onto Cylc Task status
+const status = {
+  waiting: 'waiting',
+  submitted: 'submitted',
+  succeeded: 'succeeded',
+  failed: 'failed',
+  submitFailed: 'submit-failed'
+}
+Object.freeze(status)
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply src/components/cylc/Mutation.vue]: This is probably another temporary detail, although I do like the idea of using our task icons in combination with a [stepper](https://vuetifyjs.com/en/components/steppers) for a more consistent "Cylcy" interface.

~Cylcey~
~Cylkey~
~Sylkey~
~Silkey~
Cylcy?

File: src/components/cylc/Mutation.vue [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+          mutation: gql(constructMutation(this.mutation)),
+          variables: args
+        })
+        console.log('there')
+      } catch (err) {
+        console.log(err)
+        this.status = status.submitFailed
+        return
+      }
+      console.log(result)
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply src/components/cylc/Mutation.vue]: (I should have removed these `console.log`s anyway)

File: src/utils/graphql.js [hunk 5]
```diff
@@ -120,3 +121,126 @@ export function createApolloClient (httpUrl, subscriptionClient) {
...
 }
+
+/** Walk a GraphQL type yielding all composite types on route.
+ *
+ * E.G. NonNull<List<String>> would yield:
+ *  1. NonNull
+ *  2. List
+ *  3. String
+ *
+ * @param type {Object} A type as returned by an introspction query.
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply src/utils/graphql.js]: Dammit, I had spellcheck turned on as well!

File: src/utils/graphql.js [hunk 6]
```diff
@@ -120,3 +121,126 @@ export function createApolloClient (httpUrl, subscriptionClient) {
...
+/** Return an appropriate null value for the specified type.
+ *
+ * @param type {Object} A type field as returned by an introspection query.
+ * (an object of the form {name: x, kind: y, ofType: z}).
+ * @param types {Array} An array of all types present in the schema.
+ * (optional: used to resolve InputObjectType fields).
+ *
+ * @returns {Object|Array|undefined}
+ */
+export function getNullValue (type, types) {
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply src/utils/graphql.js]: On the line bellow it does:

```js
if (!types) {
    types = []
}
```

[the remark this answers is not in the document]
Comment (kinow) [reply src/utils/graphql.js]: Yup. But default parameter values in JS work differently. It's safe to use `a=[]` in JS. As that won't re-use the same instance. https://nbviewer.jupyter.org/github/kinow/notebooks/blob/master/javascript/default-mutable-params/Default%20mutable%20params.ipynb

File: src/components/cylc/workflow/index.js [hunk 7]
```diff
@@ -130,8 +131,39 @@ const GraphWrapper = Vue.component('graph-wrapper', {
...
+    Mutations
+  },
+  mounted () {
+    const widgetElement = document.getElementById(this.widgetId)
+    widgetElement.appendChild(this.$refs[this.widgetId].$el)
+    const vm = this
+    document.getElementById(this.widgetId).addEventListener('delete:widgetcomponent', () => {
+      EventBus.$emit('delete:widget', { id: vm.widgetId })
+      vm.$destroy()
+    }, false)
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply src/components/cylc/workflow/index.js]: Think I've fixed this correctly, could you take a look at the latest commit to check I've done it right.

File: src/utils/graphql.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply src/utils/graphql.js]: Replied on the comment below. In (plain old-fashioned) JavaScript it's fine to have optional parameters, they get set to `null` if not provided. This is the JS phillosophy, if something isn't set just assume it's `null` and continue, similar to bash.

If you're running with `strict on` that might change things, but if so we aren't doing that?

File: src/utils/graphql.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply src/utils/graphql.js]: Nice.

File: src/utils/graphql.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (kinow) [reply src/utils/graphql.js]: I think `strict` is enabled by default with our current Vue setup. My IDE parses our webpack/vue-loader configuration, and when it inspected the code it added a warning to the missing value. Definitely not a blocker, so marking as resolved 👍

[the remark this answers is not in the document]
Comment (oliver-sanders) [reply src/utils/graphql.js]: (have changed BTW)

[the remark this answers is not in the document]
Comment (oliver-sanders) [reply src/utils/graphql.js]: The old approach did work though so I'm not convinced we are using `strict = true`, or perhaps webpack is getting involved?

**Comment by hjoliver** [comment]
@oliver-sanders at least my attempt passed the linter :grin: 

```
error: Operator '=' must be spaced (space-infix-ops) at src/utils/graphql.js:153:42:
  151 |  * @returns {Object|Array|undefined}
  152 |  */
> 153 | export function getNullValue (type, types=[]) {
      |                                          ^
  154 |   let ret = null
  155 |   let ofType = null
  156 |   for (const subType of iterateType(type)) {
```

**Comment by oliver-sanders** [comment]
> at least my attempt passed the linter

The subtle differences between Python and JavaScript make it really horrible to quickly switch between the two.

**Pull request merged**